### PR TITLE
CATROID-1647 Feat:  Resize and rotate object when placing in stage

### PR DESCRIPTION
**Pull request by harshsomankar123-tech**
#### Related Jira Ticket
(https://catrobat.atlassian.net/browse/CATROID-1647)

#### Summary
This PR introduces multi-touch support to the `VisualPlacementActivity`, allowing users to intuitively **resize** and **rotate** a sprite while positioning it on the stage. 

To provide clear visual feedback, a new **bounding box overlay** with corner handles has been implemented. Additionally, the IDE now automatically handles data persistence by inserting `SetSizeToBrick` and `PointInDirectionBrick` into the sprite's script upon confirmation, streamlining the creation workflow.

#### Technical Changes
* **`ResizeRotateGestureDetector.kt`**: New gesture detector for handling pinch-to-scale and two-finger rotation math.
* **`BoundingBoxOverlay.kt`**: Custom View that draws a dynamic blue outline and curved-arrow handles around the active sprite.
* **`VisualPlacementTouchListener.java`**: Updated to coordinate between single-finger dragging and multi-finger transformations.
* **`TransformationInterface.kt`**: Extends the existing interface to include scale and rotation setters.
* **`ProjectActivity.kt`**: Added logic to automatically append the relevant bricks to the script based on the visual placement results.

### Screen Recording

https://github.com/user-attachments/assets/0ef5358c-1198-42b3-b6cd-8cec4c9cae33

Undo and Rotate fun 

https://github.com/user-attachments/assets/45799534-a62a-421f-ae40-2063f2f624ee

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Branch-and-Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [x] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

[IDE-217]: https://catrobat.atlassian.net/browse/IDE-217?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/visualplacement/VisualPlacementActivity.java [hunk 1]
```diff
@@ -186,7 +203,29 @@ protected void onCreate(@Nullable Bundle savedInstanceState) {
...
+					imageView.setScaleX(appliedScale);
+					imageView.setScaleY(appliedScale);
+				}
+			}
+
+			@Override
+			public void onRotate(float rotation) {
+				VisualPlacementActivity.this.rotation = rotation;
+				if (imageView != null) {
+					imageView.setRotation(initialSpriteRotation + VisualPlacementActivity.this.rotation);
```
[the remark this answers is not in the document]
Comment (harshsomankar123-tech) [reply catroid/src/main/java/org/catrobat/catroid/visualplacement/VisualPlacementActivity.java]: Thanks! I’ve addressed it in 
d039f50

File: catroid/src/main/java/org/catrobat/catroid/ui/ProjectActivity.kt [hunk 2]
```diff
@@ -258,11 +264,24 @@ class ProjectActivity : BaseCastActivity() {
...
                 currentSprite.prependScript(startScript)
                 startScript.addBrick(placeAtBrick)
+                if (placementScale != DEFAULT_SCALE) {
+                    val sizePercent = placementScale.toDouble() *
+                        PERCENTAGE_MULTIPLIER
+                    startScript.addBrick(SetSizeToBrick(sizePercent))
+                }
+                if (placementRotation != DEFAULT_ROTATION) {
+                    val direction = placementRotation.toDouble()
+                    startScript.addBrick(PointInDirectionBrick(direction))
```
Comment (wslany) [inline catroid/src/main/java/org/catrobat/catroid/ui/ProjectActivity.kt]: [P3] `VisualPlacementActivity` returns the look's actual default direction, which is 90 degrees for a new look, but this check compares against `0.0f`. Adding a new object without rotating it therefore still appends a `PointInDirectionBrick(90)`, cluttering the generated script even though the behavior is unchanged. Compare against the Catroid default/current direction instead.

File: catroid/src/main/java/org/catrobat/catroid/visualplacement/ResizeRotateGestureDetector.kt [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+
+            MotionEvent.ACTION_MOVE ->
+                if (isTransforming && initialDistance > 0) {
+                    val scaleFactor = currentDistance / initialDistance
+                    val newScale = (cumulativeScale * scaleFactor)
+                        .coerceIn(MIN_SCALE, MAX_SCALE)
+                    listener.onScale(newScale)
+
+                    val angleDelta = currentAngle - initialAngle
+                    listener.onRotate(cumulativeRotation + angleDelta)
```
Comment (wslany) [inline catroid/src/main/java/org/catrobat/catroid/visualplacement/ResizeRotateGestureDetector.kt]: [P2] `currentAngle - initialAngle` jumps from about +179 to -179 degrees when the two fingers cross the atan2 boundary, so a tiny clockwise rotation can be reported as roughly -358 degrees. Normalize the delta into the shortest signed angle before calling `onRotate` and before accumulating `cumulativeRotation` on pointer-up.

File: catroid/src/main/java/org/catrobat/catroid/visualplacement/VisualPlacementTouchListener.java [hunk 4]
```diff
@@ -55,63 +58,112 @@ private void setMode(Mode mode) {
...
+
+	private boolean handleMultiTouch(MotionEvent event) {
+		if (resizeRotateDetector == null) {
+			return false;
+		}
+		if (event.getPointerCount() >= 2 || resizeRotateDetector.isTransforming()) {
+			boolean handled = resizeRotateDetector.onTouchEvent(event);
+			if (handled || resizeRotateDetector.isTransforming()) {
+				updateBoundingBox();
+				return true;
```
Comment (wslany) [inline catroid/src/main/java/org/catrobat/catroid/visualplacement/VisualPlacementTouchListener.java:99]: [P2] After a pinch/rotate gesture is consumed, the listener returns without updating `previousX`/`previousY` or resetting the tap/move mode. If the user lifts one finger and keeps dragging with the other, the next single-finger `ACTION_MOVE` computes its delta from the stale coordinates saved before the transform, which can jump the sprite unexpectedly. Reset the single-touch baseline when the transform ends.

File: catroid/src/main/java/org/catrobat/catroid/ui/SpriteActivity.java [hunk 5]
```diff
@@ -104,6 +104,8 @@
 import static org.catrobat.catroid.ui.SpriteActivityOnTabSelectedListenerKt.removeTabLayout;
 import static org.catrobat.catroid.ui.WebViewActivity.MEDIA_FILE_PATH;
 import static org.catrobat.catroid.visualplacement.VisualPlacementActivity.CHANGED_COORDINATES;
+import static org.catrobat.catroid.visualplacement.VisualPlacementActivity.ROTATION_BUNDLE_ARGUMENT;
```
Comment (wslany) [inline catroid/src/main/java/org/catrobat/catroid/ui/SpriteActivity.java:108]: [P3] This file is touched in this PR, but its copyright header still says `2010-2025`. The two existing visual placement Java tests touched by this PR have the same issue. For 2026 work, please update touched Catroid files to the canonical `Copyright (C) 2010-2026 The Catrobat Team` text.

**Review by wslany — CHANGES_REQUESTED** [review]
Thanks for the feature. I found a couple of interaction edge cases around rotation handling and the transition back from multi-touch to single-touch, plus two smaller cleanup items. The generated APK builds and the focused visual-placement unit tests pass locally, but I think the P2 items should be fixed before merge.

File: catroid/src/main/java/org/catrobat/catroid/ui/ProjectActivity.kt [hunk 6]
```diff
@@ -89,6 +92,10 @@ class ProjectActivity : BaseCastActivity() {
         const val SPRITE_CAMERA = 3
         const val SPRITE_OBJECT = 4
         const val SPRITE_FROM_LOCAL = 5
+
+        const val DEFAULT_SCALE = 1.0f
+        const val PERCENTAGE_MULTIPLIER = 100.0
+        const val DEFAULT_ROTATION = 0.0f
```
Comment (wslany) [inline catroid/src/main/java/org/catrobat/catroid/ui/ProjectActivity.kt]: [P3] `DEFAULT_ROTATION` is now unused after comparing against `DEGREE_UI_OFFSET`. It should be removed, or the default extra read should use a clearly named Catroid default-direction constant, so the file does not carry a misleading 0-degree default.

File: catroid/src/main/java/org/catrobat/catroid/visualplacement/VisualPlacementTouchListener.java [hunk 7]
```diff
@@ -55,63 +58,115 @@ private void setMode(Mode mode) {
...
+			}
+			previousX = event.getRawX();
+			previousY = event.getRawY();
+			setMode(Mode.MOVE);
```
Comment (wslany) [inline catroid/src/main/java/org/catrobat/catroid/visualplacement/VisualPlacementTouchListener.java]: [P2] The new reset only runs when `resizeRotateDetector.onTouchEvent(event)` returns false. For the usual `ACTION_POINTER_UP` event, `pointerCount` is still at least 2 and the detector returns true after setting `isTransforming` false, so this method returns before updating `previousX`/`previousY`. The next one-finger move can still use the stale pre-transform baseline and jump the sprite.

File: catroid/src/test/java/org/catrobat/catroid/test/visualplacement/ResizeRotateGestureDetectorTest.kt [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+    @Test
+    fun testSetCumulativeRotation() {
+        detector.cumulativeRotation = 45.0f
+        assertEquals(45.0f, detector.cumulativeRotation, DELTA)
+    }
+
+    @Test
+    fun testSingleFingerDoesNotTransform() {
+        assertFalse(detector.isTransforming)
```
Comment (wslany) [inline catroid/src/test/java/org/catrobat/catroid/test/visualplacement/ResizeRotateGestureDetectorTest.kt:130]: [P2] The follow-up commit fixes angle normalization, touch baseline reset, and default-direction brick suppression in production code, but adds no regression tests for those behaviors. These detector tests only cover helper angle/distance math and initial state, so the +/-180 degree wrap fix could regress without a failing test.

**Review by wslany — CHANGES_REQUESTED** [review]
Thanks for the quick follow-up. I still see a couple of actionable issues in the latest head, mainly around the pointer-up reset path and missing regression coverage for the review fixes.

File: catroid/src/main/java/org/catrobat/catroid/ui/ProjectActivity.kt (hunk 6 above)
Comment (harshsomankar123-tech) [reply catroid/src/main/java/org/catrobat/catroid/ui/ProjectActivity.kt]: Done. Removed DEFAULT_ROTATION and updated the default read to use DEGREE_UI_OFFSET (90.0f)

File: catroid/src/main/java/org/catrobat/catroid/visualplacement/VisualPlacementTouchListener.java (hunk 7 above)
Comment (harshsomankar123-tech) [reply catroid/src/main/java/org/catrobat/catroid/visualplacement/VisualPlacementTouchListener.java]: Good catch! I've Refactored handleMultiTouch to reset the baseline when multi-touch ends, ensuring a smooth transition back to single-finger dragging.

File: catroid/src/test/java/org/catrobat/catroid/test/visualplacement/ResizeRotateGestureDetectorTest.kt (hunk 8 above)
Comment (harshsomankar123-tech) [reply catroid/src/test/java/org/catrobat/catroid/test/visualplacement/ResizeRotateGestureDetectorTest.kt:130]: @wslany Thanks! for review. I Add New regression tests for normalizeAngle and the multi-touch transition reset have been added in ResizeRotateGestureDetectorTest.kt and VisualPlacementResizeRotateTest.kt

File: catroid/src/main/java/org/catrobat/catroid/ui/ProjectActivity.kt [hunk 9]
```diff
@@ -258,11 +264,24 @@ class ProjectActivity : BaseCastActivity() {
...
+                val currentSprite = projectManager.currentSprite ?: return
                 val startScript = StartScript()
                 currentSprite.prependScript(startScript)
                 startScript.addBrick(placeAtBrick)
+                if (placementScale != DEFAULT_SCALE) {
+                    val sizePercent = placementScale.toDouble() *
+                        PERCENTAGE_MULTIPLIER
+                    startScript.addBrick(SetSizeToBrick(sizePercent))
+                }
+                if (placementRotation != DEGREE_UI_OFFSET) {
```
Comment (wslany) [inline catroid/src/main/java/org/catrobat/catroid/ui/ProjectActivity.kt]: [P2] Please add direct regression coverage for this visual-placement result path. The focused Jacoco run covers 0/12 added executable lines in `ProjectActivity.kt`, and I could not find a test that returns `REQUEST_CODE_VISUAL_PLACEMENT` with the default `DEGREE_UI_OFFSET` rotation and asserts that no `PointInDirectionBrick` is added. Since this path creates the generated script, we should cover the default-rotation suppression and the non-default scale/rotation brick additions before merging.

**Review by wslany — CHANGES_REQUESTED** [review]
One remaining coverage request: we want full coverage for new behavior from now on. The focused unit coverage improved for the gesture/touch code, but the activity result/script-creation path still lacks direct regression coverage.
